AOC [20.01](https://adventofcode.com/2020/day/1)

Puzzle 1:
Find two numbers in a list that sum to 2020.
Their product is the answer to the question.

Solution:
1. I could use combinatics to find all the pairs of numbers and add them, but that seems overly complex.
1. Easier: put all numbers in a list and go through it.
1. For each number, look at all subsequent numbers and if they add to 2020, we've found the match.
1. Or, easier, for each number, find the target (2020-num) and just check to see if target is in the list.



In [1]:
data='''1721
979
366
299
675
1456'''

for line in data.split('\n'):
    print(line)

1721
979
366
299
675
1456


In [5]:
numlist = [int(num) for num in data.split('\n')]
for num in numlist:
    target = 2020-num
    if target in numlist:
        print(f"{num}x{target}={num*target}")
        break

1721x299=514579


In [10]:
# easy peasy -- let's get the data and test it!
with open ('2001input.txt') as f_in:
    #data = f_in.read().split('\n')
    numlist = [int(num) for num in f_in.readlines()]

for num in numlist:
    target = 2020-num
    if target in numlist:
        print(f"{num}x{target}={num*target}")
        break

447x1573=703131


703131 is correct for my data.

Part 2: using the same data, find three numbers that add to 2020.

Solution:
1. This is trickier. I think I will use the combinatics library this time to find all the possible pairs of numbers. Then I can do the same thing I did above: find the sum of each pair and look for a target.

In [14]:
from itertools import combinations
[numpair for numpair in combinations(data.split('\n'),2)]


[('1721', '979'),
 ('1721', '366'),
 ('1721', '299'),
 ('1721', '675'),
 ('1721', '1456'),
 ('979', '366'),
 ('979', '299'),
 ('979', '675'),
 ('979', '1456'),
 ('366', '299'),
 ('366', '675'),
 ('366', '1456'),
 ('299', '675'),
 ('299', '1456'),
 ('675', '1456')]

In [21]:
numlist = [int(num) for num in data.split('\n')]
numpairs = [numpair for numpair in combinations(numlist,2)]

for pair in numpairs:
    target = 2020 - (pair[0] + pair[1])
    #print(pair,target)
    if target in numlist:
        print(pair,target,pair[0]*pair[1]*target)

(979, 366) 675 241861950
(979, 675) 366 241861950
(366, 675) 979 241861950


In [22]:
# Ok, that works. Now to the big data list
with open ('2001input.txt') as f_in:
    #data = f_in.read().split('\n')
    numlist = [int(num) for num in f_in.readlines()]

numpairs = [numpair for numpair in combinations(numlist,2)]

for pair in numpairs:
    target = 2020 - (pair[0] + pair[1])
    #print(pair,target)
    if target in numlist:
        print(pair,target,pair[0]*pair[1]*target)



(930, 609) 481 272423970
(930, 481) 609 272423970
(609, 481) 930 272423970
